include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -33,6 +37,87 @@ namespace execution
 inline namespace __dpl
 {
 
+namespace __internal
+{
+
+#if _ONEDPL_PREDEFINED_POLICIES
+struct __global_instance_tag {};
+#endif
+using _Queue_factory = sycl::queue (*)();
+
+class alignas(sycl::queue) __queue_holder
+{
+    static_assert(sizeof(sycl::queue) >= sizeof(std::pair<uintptr_t, _Queue_factory>));
+    static_assert(alignof(sycl::queue) >= alignof(std::uintptr_t));
+
+    union
+    {
+        sycl::queue __q;
+        std::pair<uintptr_t, _Queue_factory> __flag_and_factory;
+    };
+
+    bool
+    __has_queue() const
+    {
+        bool res = true;
+#if _ONEDPL_PREDEFINED_POLICIES
+        // If the first size-of-pointer bytes are zeros, we consider there is no valid queue.
+        res = (__flag_and_factory.first != 0);
+        std::atomic_signal_fence(std::memory_order_acq_rel); // to prevent possible reordering due to type punning
+#endif
+        return res;
+    }
+
+  public:
+    template <typename... _Args>
+    __queue_holder(_Args... __args) : __q(std::forward<_Args>(__args)...) {}
+
+#if _ONEDPL_PREDEFINED_POLICIES
+    // The ctor for predefined policy instances does not create a queue but stores a queue factory.
+    // The first size-of-pointer bytes - the "flag" - are nullified to indicate that there is no valid queue.
+    // Then a pointer to a factory function is stored.
+    __queue_holder(__global_instance_tag, _Queue_factory __f) : __flag_and_factory(0, __f) {}
+#endif
+
+    ~__queue_holder()
+    {
+        if (__has_queue())
+            __q.~queue();
+        else
+            __flag_and_factory.second = nullptr;
+    }
+
+    // Copy and move operations have to be explicit
+    __queue_holder(const __queue_holder& __h) : __q(__h.__get_queue()) {}
+    __queue_holder(__queue_holder&& __h) : __q(__h.__get_queue()) {}
+
+    __queue_holder&
+    operator=(const __queue_holder& __h)
+    {
+        if (this != &__h)
+            __q = __h.__get_queue();
+        return *this;
+    }
+
+    __queue_holder&
+    operator=(__queue_holder&& __h)
+    {
+        if (this != &__h)
+            __q = __h.__get_queue();
```

Comment:
Looks like simple assignment, move should work a little bit faster.

Reply:
No. `__get_queue` returns an rvalue object, not a reference. Compilers should use return value optimization (RVO) techniques to not even create that object, but directly update the left side of the assignment instead. Adding `std::move`` will unfortunately prohibit RVO. as far as I understand. @rarutyun can comment more on this.

Reply:
Yes. And this mean that probably this operation should be implemented by some another way. For example - to avoid `syc::queue` creation inside `__get_queue` call.

Reply:
`__get_queue` can only return an object, because that's what the factory produces if it is used in case the holder has no valid queue.
I will think though if something reasonable can be done.

Reply:
Something like this can be done in order to move a queue in case `__h` already holds it (essentially, to move a `shared_ptr` instead of copying):
```
    __queue_holder(__queue_holder&& __h) {
        if (h.__has_queue())
            __q = std::move(__h.__g); // perhaps there should be placement new, for visibility if nothing else
        else
            __q = (__h.__flag_and_factory.second)(); // here as well
    }

    __queue_holder&
    operator=(__queue_holder&& __h)
    {
        if (this != &__h) {
            if (h.__has_queue())
                __q = std::move(__h.__g);
            else
                __q = (__h.__flag_and_factory.second)();
        }
        return *this;
    }
```
Opinions are welcome on whether it makes sense. I do not see how that can be simplified, given that the factory returns a temporary object. not a reference.

Reply:
Probably at least we may incapsulate this code into some private static function and pass `*this` and `__queue_holder&& __h` into it.

Reply:
If I'm understanding correctly, I think something like what is proposed here is a benefit.  I don't like that we are "reaching in" to the `__h` object to be able to get a return type that we can optimize appropriately.  However, I think it comes with the territory of this sort of "split" union type.  At least the specialized code still lives within this `__queue_holder` class and it does not infect other places with complexity.

I agree with Sergey's suggestion that we can consolidate the repeated code for the two move routines.

Reply:
Upon the second thought, it is simpler than the above. For move operations, we should require that the moved policy has a proper queue, because moving the predefined device policies is a horrible idea.